### PR TITLE
fix(@rjsf/antd): use tooltip for description

### PR DESCRIPTION
**Pull request by ianwith**
### Reasons for making this change

After I made this [PR](https://github.com/rjsf-team/react-jsonschema-form/pull/2644), I realized that `extra` is not the right place to put description, it is usually used for prompt message, refer to the [document of ant-design](https://ant.design/components/form/):
```
The extra prompt message. It is similar to help. Usage example: to display error message and prompt message at the same time
```

Instead, we should use `tooltip` for description. It will look like:
image

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [ ] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/master/CHANGELOG.md) with a description of the PR
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by ianwith** [comment]
@epicfaace could you check this quick fix please? We are using this library extensively in our projects😄 Thanks!

**Comment by ianwith** [comment]
@epicfaace Sure.
Putting it in `extra` will be like
image

If there is some error:
image

**Comment by ianwith** [comment]
If we put it in the `tooltip`, the description will appear when we hover over the question mark next to the title

**Comment by ianwith** [comment]
Updates on this:
1. Check `rawDescription` before using `description` to ensure we don't show question mark when there's no description
2. Upgrade `antd` dependencies to `4.7.0` for tooltip support

**Comment by epicfaace** [comment]
Thanks. Can you add this to the meeting agenda? I'd like to discuss synchronously with a couple of others (feel free to join if you're able) to see what the right step forward is. https://docs.google.com/document/d/12PjTvv21k6LIky6bNQVnsplMLLnmEuypTLQF8a-8Wss/edit

**Review by epicfaace — COMMENTED** [review]
Let's toggle this feature on only when someone puts the following in:

```js
{
  "ui:options": {
       "antd": {"descriptionLocation": "tooltip"}
   }
}
```

options:

`"descriptionLocation": "tooltip"`

`"descriptionLocation": "below"`

File: packages/antd/package.json [hunk 1]
```diff
@@ -27,16 +27,16 @@
     }
   },
   "peerDependencies": {
-    "@ant-design/icons": "^4.0.0",
+    "@ant-design/icons": "^4.7.0",
     "@rjsf/core": "^4.0.0",
-    "antd": "^4.0.0",
+    "antd": "^4.7.0",
```
Comment (epicfaace) [inline packages/antd/package.json:32]:
```suggestion
    "antd": "^4.0.0",
```

File: packages/antd/package.json [hunk 2]
```diff
@@ -27,16 +27,16 @@
     }
   },
   "peerDependencies": {
-    "@ant-design/icons": "^4.0.0",
+    "@ant-design/icons": "^4.7.0",
```
Comment (epicfaace) [inline packages/antd/package.json:30]:
```suggestion
    "@ant-design/icons": "^4.0.0",
```

File: packages/antd/README.md [hunk 3]
```diff
@@ -56,8 +56,8 @@ Ant Design theme, fields and widgets for `react-jsonschema-form`.
 
 ### Prerequisites
 
-- `antd >= 4.0.0`
-- `@ant-design/icons >= 4.0.0`
+- `antd >= 4.7.0`
+- `@ant-design/icons >= 4.7.0`
```
Comment (epicfaace) [inline packages/antd/README.md:60]:
```suggestion
- `@ant-design/icons >= 4.0.0`
```

File: packages/antd/README.md [hunk 4]
```diff
@@ -56,8 +56,8 @@ Ant Design theme, fields and widgets for `react-jsonschema-form`.
 
 ### Prerequisites
 
-- `antd >= 4.0.0`
-- `@ant-design/icons >= 4.0.0`
+- `antd >= 4.7.0`
```
Comment (epicfaace) [inline packages/antd/README.md:59]:
```suggestion
- `antd >= 4.0.0`
```

File: packages/antd/README.md (hunk 3 above)
Comment (epicfaace) [inline packages/antd/README.md:60]: @ianwith if you could make another PR that bumps the peer dependency of antd to 4.7, we can then merge that in once we release the next major version of rjsf.

**Comment by ianwith** [comment]
@epicfaace Is there a way to put a global `ui:options` so that we don't need to write it in each field, or could we put `descriptionLocation` in formContext?

**Comment by epicfaace** [comment]
There's no way to do a global ui:options. Maybe you could try doing
something with formContext?

--
Ashwin Ramaswami

**Comment by ianwith** [comment]
@epicfaace I reverted `antd` upgrade and added `descriptionLocation` in `formContext`, please review. Thanks!

File: docs/api-reference/themes/antd/uiSchema.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+```jsx
+<Form
+  formContext={{
+    "antd": {
+      "descriptionLocation": "tooltip"
+    }
+    // other props...
+  }}
+/>
+```
```
Comment (epicfaace) [inline docs/api-reference/themes/antd/uiSchema.md:22]: can you also move the `readonlyAsDisabled ` documentation from https://react-jsonschema-form.readthedocs.io/en/latest/api-reference/form-props/#formcontext to here? And replace that with "see [page X] for formContext customizations for the antd theme"

File: packages/antd/test/__snapshots__/Array.test.js.snap [hunk 6]
```diff
@@ -415,21 +434,16 @@ exports[`array fields fixed array 1`] = `
                             className="ant-input-number-input-wrap"
                           >
                             <input
-                              aria-valuemin={-9007199254740991}
+                              aria-valuenow={null}
```
Comment (epicfaace) [inline packages/antd/test/__snapshots__/Array.test.js.snap:437]: do you know why this change happened? is this because of the antd upgrade?

File: packages/antd/src/templates/FieldTemplate/index.js [hunk 7]
```diff
@@ -63,9 +67,10 @@ const FieldTemplate = ({
       ) : (
         <Form.Item
           colon={colon}
-          extra={description}
+          tooltip={descriptionLocation === 'tooltip' && (!!rawDescription && description)}
+          extra={descriptionLocation === 'below' && description}
           hasFeedback={schema.type !== 'array' && schema.type !== 'object'}
-          help={(!!rawHelp && help) || (!!rawErrors && renderFieldErrors())}
+          help={(rawHelp ? help : undefined) || (rawErrors ? renderFieldErrors() : undefined)}
```
Comment (epicfaace) [inline packages/antd/src/templates/FieldTemplate/index.js:73]: revert this? since we use the same pattern (`!!rawDescription && description`) above

Comment (ianwith) [reply packages/antd/src/templates/FieldTemplate/index.js:73]: The problem of `!!rawHelp && help` is that it's value could be `false` which will result in a blank area being displayed in a new version of in `antd` because `antd` checks the presence of help message by `help !== undefined || help !== null`.

Comment (ianwith) [reply packages/antd/src/templates/FieldTemplate/index.js:73]: @epicfaace do you think we can change `(!!rawDescription && description)` to `rawDescription ? description : undefined`

File: packages/antd/test/__snapshots__/Array.test.js.snap (hunk 6 above)
Comment (ianwith) [reply packages/antd/test/__snapshots__/Array.test.js.snap:437]: I'm not quite sure since I reverted the upgrade and re-installed all the packages.🤔

File: packages/antd/src/templates/FieldTemplate/index.js (hunk 7 above)
Comment (ianwith) [reply packages/antd/src/templates/FieldTemplate/index.js:84]: @epicfaace updated

**Comment by ianwith** [comment]
Hi, @epicfaace , I've updated this based on your comments, does it looks good to you?

**Comment by epicfaace** [comment]
Hi, if possible, can we do a last synchronous review of this during the weekly meeting time? Can you add it to the schedule? https://github.com/rjsf-team/react-jsonschema-form/issues/2677

Thanks!

**Comment by ianwith** [comment]
Sure! added to the docs.

File: packages/antd/src/templates/FieldTemplate/index.js [hunk 8]
```diff
@@ -34,8 +35,24 @@ const FieldTemplate = ({
...
+    antd
   } = formContext;
 
+  const descriptionLocation= antd?.descriptionLocation || 'below';
+
+  const descriptionProps = {};
+
+  switch (descriptionLocation) {
+    case 'tooltip':
+      descriptionProps.tooltip = rawDescription ? description : undefined;
```
Comment (epicfaace) [inline packages/antd/src/templates/FieldTemplate/index.js:47]: I think this should work. let's check and if this works, let's do this:
```suggestion
      descriptionProps.tooltip = description;
```

otherwise add a comment for why `rawDescription ? description : undefined` is needed

File: packages/antd/src/templates/FieldTemplate/index.js (hunk 7 above)
Comment (epicfaace) [reply packages/antd/src/templates/FieldTemplate/index.js:84]: > The problem of `!!rawHelp && help` is that it's value could be `false` which will result in a blank area being displayed in a new version of in `antd` because `antd` checks the presence of help message by `help !== undefined || help !== null`.

Let's add a few tests:
- rawDescription is false
- rawHelp is false
- rawErrors is false

All tests should ensure that a blank area is not displayed.

Comment (ianwith) [reply packages/antd/src/templates/FieldTemplate/index.js:84]: @epicfaace I added test for `ui:help` https://github.com/rjsf-team/react-jsonschema-form/pull/2740/commits/c7abf91977ea51ae6b9705a85b6cfe7ab73f8450

I think we don't need to add test for `false` case because all test without `ui:help` passed, indicating that this change had no effect on the output.

The reason I made this change is only to remove blank area for newer antd version.

File: packages/antd/src/templates/FieldTemplate/index.js (hunk 8 above)
Comment (ianwith) [reply packages/antd/src/templates/FieldTemplate/index.js:47]: updated `extra` prop with `rawDescription ? description : undefined`

**Comment by heath-freenome** [comment]
@ianwith I'm willing to approve this after you fix the conflicts resulting from the v5 beta changes... Thanks!

**Comment by miguelgrc** [comment]
Hey, is this still being worked on? It would be a very useful feature for me as well

**Comment by heath-freenome** [comment]
@ianwith @miguelgrc Would one of you like to complete this PR?

**Comment by heath-freenome** [comment]
closing as this has been implemented in #3524
